### PR TITLE
Use exceptions to communicate failure over dbus

### PR DESCRIPTION
**Pull request by DunderRoffe**
Remove bool success returns in DBus API

Add support for converting SoftwareContainerError in agent to
`org.freedesktop.DBus.Error.Failed` response on DBus API.

Refactor servicetests to handle DBusException instead of
checking success value.

Refactor agent componenttests to handle SoftwareContainerError
instead of checking success value.

Signed-off-by: Viktor Sjölind <viktor.sjolind@pelagicore.com>

### CODE REVIEW

File: agent/component-test/softwarecontaineragent_unittest.cpp [hunk 1]
```diff
@@ -111,19 +112,15 @@ class SoftwareContainerAgentTest: public ::testing::Test
 };
 
 TEST_F(SoftwareContainerAgentTest, CreatAndCheckContainer) {
-    ContainerID id;
-    bool success = sca->createContainer(valid_config, id);
-    ASSERT_TRUE(success);
-    ASSERT_TRUE(sca->getContainer(id) != nullptr);
+    ContainerID id = sca->createContainer(valid_config);
```
Comment (kursatkobya) [inline agent/component-test/softwarecontaineragent_unittest.cpp]: Maybe using like `ASSERT_NO_THROW(id = sca->createContainer(valid_config));` would be better

File: agent/component-test/softwarecontaineragent_unittest.cpp [hunk 2]
```diff
@@ -111,19 +112,15 @@ class SoftwareContainerAgentTest: public ::testing::Test
...
+    ASSERT_NO_THROW(sca->getContainer(id));
 }
 
 TEST_F(SoftwareContainerAgentTest, DeleteContainer) {
-    ContainerID id;
-    bool success = sca->createContainer(valid_config, id);
-    ASSERT_TRUE(success);
+    ContainerID id = sca->createContainer(valid_config);
```
Comment (kursatkobya) [inline agent/component-test/softwarecontaineragent_unittest.cpp]: `ASSERT_NO_THROW(id = sca->createContainer(valid_config));` could be better

File: agent/component-test/softwarecontaineragent_unittest.cpp [hunk 3]
```diff
@@ -140,119 +137,106 @@ TEST_F(SoftwareContainerAgentTest, CreateContainerWithConf) {
...
+    ASSERT_NO_THROW(sca->resumeContainer(id));
 }
 
 // Freeze an already frozen container
 TEST_F(SoftwareContainerAgentTest, FreezeFrozenContainer) {
-    ContainerID id;
-    bool success = sca->createContainer(valid_config, id);
-    ASSERT_TRUE(success);
-    ASSERT_TRUE(sca->getContainer(id) != nullptr);
+    ContainerID id = sca->createContainer(valid_config);
```
Comment (kursatkobya) [inline agent/component-test/softwarecontaineragent_unittest.cpp]: using `ASSERT_NO_THROW` would be beter

Comment (kursatkobya) [inline agent/component-test/softwarecontaineragent_unittest.cpp]: here as well

File: agent/component-test/softwarecontaineragent_unittest.cpp [hunk 4]
```diff
@@ -140,119 +137,106 @@ TEST_F(SoftwareContainerAgentTest, CreateContainerWithConf) {
...
+    ASSERT_NO_THROW(sca->suspendContainer(id));
 }
 
 // Freeze an already frozen container, and then resume it
 TEST_F(SoftwareContainerAgentTest, DoubleFreezeContainerAndThaw) {
-    ContainerID id;
-    bool success = sca->createContainer(valid_config, id);
-    ASSERT_TRUE(success);
-    ASSERT_TRUE(sca->getContainer(id));
+    ContainerID id = sca->createContainer(valid_config);
```
Comment (kursatkobya) [inline agent/component-test/softwarecontaineragent_unittest.cpp]: and here

File: agent/component-test/softwarecontaineragent_unittest.cpp [hunk 5]
```diff
@@ -140,119 +137,106 @@ TEST_F(SoftwareContainerAgentTest, CreateContainerWithConf) {
...
+    ASSERT_NO_THROW(sca->resumeContainer(id));
 }
 
 // Double suspend and then double resume
 TEST_F(SoftwareContainerAgentTest, DoubleFreezeAndDoubleThawContainer) {
-    ContainerID id;
-    bool success = sca->createContainer(valid_config, id);
-    ASSERT_TRUE(success);
-    ASSERT_TRUE(sca->getContainer(id) != nullptr);
+    ContainerID id = sca->createContainer(valid_config);
```
Comment (kursatkobya) [inline agent/component-test/softwarecontaineragent_unittest.cpp]: here

File: agent/component-test/softwarecontaineragent_unittest.cpp [hunk 6]
```diff
@@ -140,119 +137,106 @@ TEST_F(SoftwareContainerAgentTest, CreateContainerWithConf) {
...
+    ASSERT_NO_THROW(sca->resumeContainer(id));
 }
 
 // Make sure you can still shutdown a frozen container
 TEST_F(SoftwareContainerAgentTest, ShutdownFrozenContainer) {
-    ContainerID id;
-    bool success = sca->createContainer(valid_config, id);
-    ASSERT_TRUE(success);
-    ASSERT_TRUE(sca->getContainer(id) != nullptr);
+    ContainerID id = sca->createContainer(valid_config);
```
Comment (kursatkobya) [inline agent/component-test/softwarecontaineragent_unittest.cpp]: h

File: agent/src/softwarecontaineragent.cpp [hunk 7]
```diff
@@ -119,94 +120,113 @@ bool SoftwareContainerAgent::triggerPreload()
...
+    }
+    
+    if (containerID < 0) {
+        std::string errorMessage("Invalid Container ID: "
+                                + std::to_string(containerID)
+                                + ". ID can not be negative");
+        log_error() << errorMessage;
+        throw SoftwareContainerError(errorMessage);
+    }
+    if (1 > m_containers.count(containerID)) {
```
Comment (kursatkobya) [inline agent/src/softwarecontaineragent.cpp:137]: maybe it could be better to use i`f (!m_containers.count(containerID)) ` 
since it can be only 1 or 0, < operator doesn't seem relevant to me

File: agent/src/softwarecontaineragent.cpp [hunk 8]
```diff
@@ -119,94 +120,113 @@ bool SoftwareContainerAgent::triggerPreload()
...
 {
-    return (containerID < INT32_MAX)
-        && (containerID >= 0)
-        && (1 == m_containers.count(containerID));
+    if (containerID >= INT32_MAX) {
+        std::string errorMessage("Invalid Container ID: "
+                                + std::to_string(containerID)
+                                + ". ID can not be greater than INT32");
+        log_error() << errorMessage;
+        throw SoftwareContainerError(errorMessage);
```
Comment (kursatkobya) [inline agent/src/softwarecontaineragent.cpp:127]: I think the error message is going to be written to output at the place SoftwareContainerError is caught. So that can create duplicate error messages.  If this is true maybe it would be sensible to remover logs before throwing errors.

File: agent/src/softwarecontaineragent.cpp (hunk 7 above)
Comment (DunderRoffe) [reply agent/src/softwarecontaineragent.cpp:137]: I do not like using `0` as `false` since they are fundamentally different types. Could go for `0 == m_containers.count(containerID)` but then negative numbers would be allowed. Although, negative numbers are not possible anyway but it still feels nice :)

Comment (kursatkobya) [reply agent/src/softwarecontaineragent.cpp:137]: `(0 == m_containers.count(containerID))` make sense as well :)

File: agent/src/softwarecontaineragent.cpp (hunk 8 above)
Comment (DunderRoffe) [reply agent/src/softwarecontaineragent.cpp:127]: I think we should try to log as quickly as possible when something happens to lessen the time spent debugging. You could argue that this can be done as well by tracing the exception, but that is not how we handle exceptions AFAIK.

File: agent/src/softwarecontaineragent.cpp [hunk 9]
```diff
@@ -316,106 +338,88 @@ bool SoftwareContainerAgent::execute(ContainerID containerID,
...
-bool SoftwareContainerAgent::shutdownContainer(ContainerID containerID)
+void SoftwareContainerAgent::shutdownContainer(ContainerID containerID)
 {
     profilefunction("shutdownContainerFunction");
     if (!m_shutdownContainers) {
-        log_info() << "Not shutting down containers, so not shutting down " << containerID;
-        return false;
+        std::string errorMessage("Not shutting down containers, so not shutting down ID: " + std::to_string(containerID));
+        log_error() << errorMessage;
+        throw SoftwareContainerError(errorMessage);
```
Comment (tobsan) [inline agent/src/softwarecontaineragent.cpp]: This should not be an error imo, this is just a fact. Just returning would suffice here.

**Review by tobsan — CHANGES_REQUESTED** [review]
Looks good, only had one comment.

**Review by tobsan — APPROVED** [review]
